### ISSUE TITLE
State object returns from PlayCookieStore as String

### ISSUE THREAD

**Issue by entropicdrifter** [pac4j/play-pac4j#167]
I'm working on a PR to fix this issue and any others that crop up in the debugging process. Currently it looks like I'll need to check during `get()` calls if `key.contentEquals("oidcStateAttribute")` and if it does, cast the object to a String, then wrap it in a State constructor before returning it.

**Comment by leleuj** [comment]
This is very strange: objects should be restored with their original type.

I'm wondering if we should not implement a different algorithm and only use the JwtGenerator for the user profiles.
Something like:

- if the value is a String, save it as "#str" + value
- if the key is the user profiles, generate a JWT and save it as "#jwt" + jwt
- otherwise, serialize the value and save it as "#ser" + serialized value

**Comment by entropicdrifter** [comment]
Okay, so I'm working on implementing the solution you've suggested. I have some questions:
is every (non-profile) Object that we'll be working with Serializable?
in cases where we get a map of user profiles (for multi-profile sessions) should I make a JWT of the whole map, or does it make more sense to make a JWT of each profile, then serialize the whole map?
Other than those things, I think I fully understand how to implement that design.

**Comment by leleuj** [comment]
I guess so. Every object put in the session should be serializable.

Good point: we have a map of profiles and not only one user profile, so we need to make a JWT out of the whole map (it should be easier that way).

**Comment by leleuj** [comment]
@entropicdrifter I'd like to release play-pac4j v2.6.2 before starting version 3 (based on pac4j 2). Do you think you can finish that this week or should I revert your changes and wait for a brand new PR?

**Comment by entropicdrifter** [comment]
You should revert it. I think I might have a way to finish it, but if I'm right it would mean you can't use the `PlayCookieStore` and the `SecurityFilter` at the same time.

**Comment by entropicdrifter** [comment]
If you'd like me to make a PR with the solution I have right now, I can. I could spend the rest of the week making it prettier for release and adding documentation talking about its limitations.

**Comment by leleuj** [comment]
That's strange: why the `PlayCookieStore` would not work with the `SecurityFilter`?

Let's try to finish that this week and at worst, I'll remove it next week before the release if things are not quite satisfactory.

**Comment by leleuj** [comment]
Where are we on this? Still planning to work on this?

**Comment by entropicdrifter** [comment]
Yes, sorry, my main project has had me pretty busy the last few months. I do still plan to figure this out.

**Comment by leleuj** [comment]
No problem. I do some cleaning sometimes, let's keep it open then ;-)

**Comment by leleuj** [comment]
What's the status on this issue?

**Comment by vidma** [comment]
what's the status on this?

**Comment by vidma** [comment]
if the issue is still present I assume this is the issue with JWTGenerator deserialization.
maybe we can live as is if that works with most common auth methods.

I'll try to test it with Forms & SAMLv2.
